// socket.io-request/lib/main.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _error = require('./error');

var _combineMiddlewares = require('combine-middlewares');

var _combineMiddlewares2 = _interopRequireDefault(_combineMiddlewares);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _toConsumableArray(arr) { if (Array.isArray(arr)) { for (var i = 0, arr2 = Array(arr.length); i < arr.length; i++) { arr2[i] = arr[i]; } return arr2; } else { return Array.from(arr); } }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

var SocketIORequest = function () {
  function SocketIORequest(io) {
    var options = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : {};

    _classCallCheck(this, SocketIORequest);

    this.io = io;
    this.options = Object.assign({
      event: 'socket.io-request',
      timeout: 90000
    }, options);
  }

  _createClass(SocketIORequest, [{
    key: 'request',
    value: function request(method, data) {
      var _this = this;

      if (typeof method !== 'string') throw new Error('argument "method" is missing');

      return new Promise(function (resolve, reject) {
        _this.io.emit(_this.options.event, { method: method, data: data }, function (res) {
          clearTimeout(timeout);
          _this.io.removeListener('disconnect', onDisconnect);
          if (res.error) return reject((0, _error.convertObjectToError)(res.error));
          resolve(res.data);
        });

        var onDisconnect = function onDisconnect() {
          clearTimeout(timeout);
          reject(new _error.SocketIOError('disconnect'));
        };

        var timeout = setTimeout(function () {
          _this.io.removeListener('disconnect', onDisconnect);
          reject(new _error.TimeoutError('exceeded ' + _this.options.timeout + ' (msec)'));
        }, _this.options.timeout);

        _this.io.once('disconnect', onDisconnect);
      });
    }
  }, {
    key: 'response',
    value: function response(method) {
      if (typeof method !== 'string') throw new Error('argument "method" is missing');

      for (var _len = arguments.length, middlewares = Array(_len > 1 ? _len - 1 : 0), _key = 1; _key < _len; _key++) {
        middlewares[_key - 1] = arguments[_key];
      }

      if (middlewares.find(function (m) {
        return typeof m !== 'function';
      })) {
        throw new Error('"middlewares" must be a function');
      }
      var combined = _combineMiddlewares2.default.apply(undefined, _toConsumableArray(middlewares.concat()));
      this.io.on(this.options.event, function (req, ack) {
        if (req.method !== method) return;
        var res = function res(data) {
          return ack({ data: data });
        };
        res.error = function (err) {
          return ack({ error: (0, _error.convertErrorToObject)(err) });
        };
        combined(req.data, res);
      });
    }
  }]);

  return SocketIORequest;
}();

exports.default = SocketIORequest;
